core/src/acc_text/nlg/gf/grammar/impl.clj

```diff
@@ -89,36 +93,34 @@
   (let [function-concept (some (fn [[role concept]]
                                  (when (= :function role) concept))
                                (zipmap (map ::sg/role relations) children))
-        role-map         (reduce (fn [m [{role ::sg/role {attr-name ::sg/name} ::sg/attributes} concept]]
-                                   (cond-> m
-                                     (and (not= :function role)
-                                          (some? attr-name)) (assoc (str/lower-case attr-name) (concept->name concept))))
-                                 {}
-                                 (zipmap relations children))]
+        role-map (reduce (fn [m [{role ::sg/role {attr-name ::sg/name} ::sg/attributes} concept]]
+                           (cond-> m
+                                   (and (not= :function role)
+                                        (some? attr-name)) (assoc (str/lower-case attr-name) (concept->name concept))))
+                         {}
+                         (zipmap relations children))]
     {:name   (concept->name concept)
      :params (map concept->name children)
      :body   (variants
                (for [syntax (->> (keyword value) (get amr) (:frames) (map :syntax))]
                  (interpose {:type  :operator
                              :value "++"}
-
-                            (for [{value :value pos :pos role :role} syntax]
-                              (let [literal-val {:type :literal :value value}]
-                                (cond
-                                  (and (some? role)
-                                       (contains? role-map (str/lower-case role))) {:type  :function
-                                                                                    :value (get role-map (str/lower-case role))}
-                                  (some? role)               {:type  :literal
-                                                              :value (format "{{%s}}" role)}
-                                  (and (some? function-concept)
-                                       (= pos :VERB))       {:type  :function
-                                                             :value (concept->name function-concept)}
-                                  (= pos :ADP)              literal-val
-                                  ;;FIXME ideally we should not have entries which are not explicitly
-                                  ;;specified via POS or other means
```

Comment:
Only two possibilities for remaining conditions - when value is defined, we display it, otherwise, we put a placeholder '{{...}}'. And if placeholders appear in wrong places it points to AMR definition being invalid. I believe the logic is fine here.

Reply:
You can have multiple roles inside AMR, but some can be optional. If user doesn't give optional role, it should simply don't generate that sentence. 

Eg. maybe there's sentence related to location, but CSV doesn't contain location field, or maybe some rows just doesn't have location. Now we would get sentences with `{{...}}` inside which doesn't look that good

Reply:
However, we can simply filter out sentences containing `{{...}}` when returning result ;)